osquery/management.py

```diff
@@ -21,6 +25,90 @@
 from osquery.extension_client import ExtensionClient, DEFAULT_SOCKET_PATH
 from osquery.extension_manager import ExtensionManager
 
+DARWIN_BINARY_PATH = "/usr/local/bin/osqueryd"
+LINUX_BINARY_PATH = "/usr/bin/osqueryd"
+
+class SpawnInstance(object):
+    """Spawn a standalone osquery instance"""
+
+    '''The osquery process instance.'''
+    instance = None
+    '''The extension client connection attached to the instance.'''
+    connection = None
+    _socket = None
+
+    def __init__(self, path=None):
+        """
+        Keyword arguments:
+        path -- the path to and osqueryd binary to spawn
+        """
+        if path is None:
+            # Darwin is special and must have binaries installed in /usr/local.
+            if sys.platform == "darwin":
+                self.path = DARWIN_BINARY_PATH
+            else:
+                self.path = LINUX_BINARY_PATH
+        else:
+            self.path = path
+        self._socket = tempfile.mkstemp(prefix="pyosqsock")
+        self._pidfile = tempfile.mkstemp(prefix="pyosqpid")
+        with open(self._pidfile[1], "w") as fh:
+            fh.write("100000")
+        self._dbpath = tempfile.mkdtemp(prefix="pyoqsdb")
+
+    def __del__(self):
+        if self.connection is not None:
+            self.connection.close()
+        if self.instance is not None:
+            self.instance.kill()
+        # Todo: remove socket, pidfile, and database
+
+    def open(self, timeout=2, interval=0.01):
+        """
+        Start the instance process and open an extension client
+
+        Keyword arguments:
+        timeout -- maximum number of seconds to wait for client
+        interval -- seconds between client open attempts
+        """
+        proc = [
+            self.path,
+            "--extensions_socket",
+            self._socket[1],
+            "--database_path",
+            self._dbpath,
+            "--pidfile",
+            self._pidfile[1]
+        ]
+        self.instance = subprocess.Popen(proc,
+            stdin=subprocess.PIPE, stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE)
+        self.connection = ExtensionClient(path=self._socket[1])
+        if not self.is_running():
+            return False
+
+        # Attempt to open the extension client.
+        delay = 0
+        while delay < timeout:
+            try:
+                self.connection.open()
+                return True
+            except:
+                time.sleep(interval)
+                delay += interval
+        return False
+
+    def is_running(self):
+        """Check if the instance has spawned."""
+        if self.instance is None:
+            return False
+        return self.instance.poll() is None
```

Comment:
Right, `.poll()` will return the `returncode` of the process when it's a zombie, otherwise it will return `None` if the process is running or has been evicted/waited by it's parent or `init`.